[OTX] Add warmstart for classification

sungchul1 opened the pull request:
This PR includes
* Warmstart for classification recipe
* Modify learning rate max value to use larger learning rate (0.45 for warmstart) than 0.1
* Add `SELFSUPERVISED` in `algo_backend.train_type.options` in `otx/algorithms/classification/configs/configuration.yaml`
* Enable to handle iter runner
* Disable validation during warmstart
* Enable to only use train set without label
* Disable evaluation after warmstart (there is an error because warmstart task doesn't have validation set)

Integration test : `tests/integration/cli/classification/test_classification.py`

FYI.
This PR has dependency with https://github.com/openvinotoolkit/model_preparation_algorithm/pull/108.

TODOs
- [x] Refactoring
- [x] Docstring

sungchul1 committed:
Add warmstart data pipeline

sungchul1 committed:
update BYOL for warmstart

sungchul1 committed:
update backend for warmstart

sungchul1 committed:
update cfgs

sungchul1 committed:
modify duplicated pipelines

sungchul1 committed:
inherit MPAClsDataset

sungchul1 committed:
improve data pipeline speed

sungchul1 committed:
update configs

sungchul1 committed:
enable to update max_iters for iter runner

sungchul1 committed:
disable evaluation when there are no val & test dataset

sungchul1 committed:
add integration test

jaegukhyun commented on otx/algorithms/classification/adapters/mmcls/data/datasets.py:
```diff
@@ -411,3 +412,48 @@ def evaluate(
+    def __getitem__(self, index: int):
+        """Get item from dataset."""
+        dataset = self.otx_dataset
+        item = dataset[index]
```
How about item = self.otx_dataset[index]?

jaegukhyun commented on otx/algorithms/classification/adapters/mmcls/models/heads/contrastive_head.py:
```diff
@@ -0,0 +1,54 @@
+# pylint: disable=missing-module-docstring, unused-argument
```
Adding short description would be more propriate than disable pylint

sungchul1 replied:
Done.

jaegukhyun commented on otx/algorithms/classification/configs/efficientnet_b0_cls_warmstart/model.py:
```diff
@@ -0,0 +1,41 @@
+    )
+)
+
+custom_hooks = [
```
We normally locate custom hooks in recipes. If warmstart needs this hook regardless models, than this could be added to warmstart recipe

sungchul1 replied:
Done.

jaegukhyun commented on otx/algorithms/common/tasks/training_base.py:
```diff
@@ -101,15 +101,16 @@ def _run_task(self, stage_module, mode=None, dataset=None, **kwargs):
-            train_data_cfg["data_classes"] = data_classes
-            new_classes = np.setdiff1d(data_classes, model_classes).tolist()
-            train_data_cfg["new_classes"] = new_classes
+        if not self._warmstart:
```
Could you explain why warmstart avoid this block?

sungchul1 replied:
Because warmstart use dataset without label, we don't need to set classes (it is unnecessary.).
But these classes will be ignored when setting `SelfSLDataset`, it's okay passing it.
If you think it is better to remove if statement, I'll fix it.

sungchul1 replied:
I removed this block for compatibility with warmstart for semantic segmentation.

jaegukhyun reviewed (changes requested):
Thank you for your hard work. It is a big PR indeed, there are several things we have to discuss
Before we start, I want to recommend to check pre-commit issue in this PR, and also recommend to follow pylint except inevitable cases for more unified and informative code.

And the followed things should be discussed with OTX TF 
1. Is it appropriate to generate separate folder for sefl-sl? Actually we have agreed on one template for one model policy
2. How we test self-sl properly? Selfsl normally requires lots of epochs to train, however it is not appropriate for test because it requires tons of time.
3. Can we implement more scalable code for warmstart? There are many if else statement for warmstart, it doesn't seem scalable. After discussing about what is different between warmstart algorithm and normal incremental or semi-supervised learning, we can agree on how to divide warmstart stage more appropriatly.

sungchul1 committed:
pylint & complete docstring

sungchul1 committed:
move custom_hooks into warmstart recipe

sungchul1 commented:
@jaegukhyun Thanks for reviewing.
1. It would be better to decide it after discussion.
2. Self-sl consumes lots of resources (time, memory, ...) as you mentioned, but warmstart is conducted with small epochs (e.g. 10 epochs) and it doesn't take much time than we expected. But I have a concern about different input size between experiment and this PR (I mentioned in https://github.com/openvinotoolkit/model_preparation_algorithm/pull/108).
3. I agree. If there is better ways to be scalable we can apply them.

sungchul1 committed:
pylint

sungchul1 committed:
remove if statement for compatibility with warmstart for semantic segmentation

sungchul1 commented on otx/algorithms/classification/tasks/inference.py:
```diff
@@ -306,16 +311,21 @@ def _init_recipe_hparam(self) -> dict:
             ),
-            runner=ConfigDict(max_epochs=int(self._hyperparams.learning_parameters.num_iters)),
         )
+        if self._warmstart:
```
set specific runner

sungchul1 commented on otx/algorithms/classification/tasks/inference.py:
```diff
@@ -345,6 +355,9 @@ def _init_recipe(self):
             recipe = os.path.join(recipe_root, "incremental.yaml")
 
+        if train_type == TrainType.SELFSUPERVISED:
+            recipe = os.path.join(recipe_root, "warmstart.yaml")
```
warmstart -> selfsl?

sungchul1 commented on otx/algorithms/classification/adapters/mmcls/models/pretrainers/byol.py:
```diff
@@ -0,0 +1,196 @@
+            Default: 0.996.
+    """
+
+    def __init__(
```
use state_dict_hook

sungchul1 commented on otx/algorithms/common/adapters/mmcv/hooks.py:
```diff
@@ -567,3 +568,78 @@ def after_train_iter(self, runner: BaseRunner):
+
+
+@HOOKS.register_module()
+class MomentumUpdateHook(Hook):
```
-> EMAMomentumUpdateHook

sungchul1 commented on tests/integration/cli/classification/test_classification.py:
```diff
@@ -33,6 +34,18 @@
+    "train_params": [
+        "params",
+        "--learning_parameters.num_iters",
+        "100",
```
-> 10

sungchul1 commented on tests/integration/cli/classification/test_classification.py:
```diff
@@ -94,8 +107,42 @@ def tmp_dir_path():
+                else:
+                    to_save_data_args["data"][mode][property] = None
+        yaml.dump(to_save_data_args, open("./data.yaml", "w"), default_flow_style=False)
+        args_w["--data"] = "./data.yaml"
```
- remove data.yaml after test

sungchul1 committed:
update recipe structure

Still remain template_experimental.yaml for current version.
It will be removed when the way the template is loaded is updated.

sungchul1 committed:
replace `warmstart` with `selfsl` & update runner condition

sungchul1 committed:
MomentumUpdateHook -> EMAMomentumUpdateHook

sungchul1 committed:
remove duplicated transforms (RandomFlip, Solarize)

sungchul1 committed:
add state_dict_hook to save online_backbone only

sungchul1 committed:
update templates

wonjuleee reviewed (approved):
I think that the only the part using pipeline1 and pipeline2 would be confusing and others are fine. how about specifying the view0 and view1 as other Self-SLs do?

sungchul2 commented:
> I think that the only the part using pipeline1 and pipeline2 would be confusing and others are fine. how about specifying the view0 and view1 as other Self-SLs do?
